### PR TITLE
fixed: SUSE / Fedora / CentOS / RHEL package dependencies and proper %if

### PR DESCRIPTION
**Pull request by mmohring**
Changes proposed in this pull request:

* Fix Buildtime and Runtime dependencies for CentOS / RHEL / Fedora
* Allow to generate OBS packages for Fedora / CentOS / RHEL with correct runtime dependencies
* Generates python 2 only package on CentOS / RHEL which have no python3
* Kiwi package is then able to be run in OBS for Fedora / RHEL / CentOS targets

### CODE REVIEW

**Comment by mmohring** [comment]
This should fix the package .spec and together with #425 also the python2 / python3 Makefile issues.

File: package/python-kiwi-spec-template [hunk 1]
```diff
@@ -231,14 +278,14 @@ outside of the scope of kiwi appliance building.
 %ifarch %ix86 x86_64
 %package -n kiwi-pxeboot
 Summary:        KIWI - PXE boot structure
+License:        GPL-3.0+
+Group:          System/Management
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: These tags do not need to be repeated from the main header, as they are inherited during the package build.

File: package/python-kiwi-spec-template [hunk 2]
```diff
@@ -66,11 +66,18 @@
 %define distro rhel-07.0
 %endif
 
+# Fedora
+# use the rhel templates for CentOS, too
+%if 0%{?fedora_version}
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Please use `%fedora` instead of `%fedora_version`.

File: package/python-kiwi-spec-template [hunk 3]
```diff
@@ -66,11 +66,18 @@
...
+%define distro fedora-25.0
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Use `fedora-%{fedora}.0` to ensure the version is selected dynamically.

File: package/python-kiwi-spec-template [hunk 4]
```diff
@@ -79,13 +86,20 @@ Source:         %{name}.tar.gz
...
+BuildRequires:  python3-devel
+BuildRequires:  python3-setuptools
+%endif
+%if 0%{?suse_version}
 BuildRequires:  fdupes
-BuildRequires:  update-alternatives
 BuildRequires:  shadow
+BuildRequires:  update-alternatives
+%endif
+%if 0%{?fedora_version} || 0%{?rhel_version} || 0%{?centos_version}
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Again, please use `%{fedora}`.

In addition, `%rhel_version` and `%centos_version` can be combined into a single `%{rhel}` check.

File: package/python-kiwi-spec-template [hunk 5]
```diff
@@ -95,57 +109,74 @@ and cloud systems like Xen, KVM, VMware, EC2 and more.
 # python2-kiwi
 %package -n python2-kiwi
 Summary:        KIWI - Appliance Builder Next Generation
+License:        GPL-3.0+
 Group:          Development/Languages/Python
-Provides:       python-kiwi = %{version}-%{release}
-Provides:       kiwi-image:tbz
+Provides:       kiwi
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Please don't do this, as this makes it difficult to have only one version installed when someone wants to install `kiwi`.

Also, I'm fairly certain that SUSE doesn't want to do this yet, so you may need to undo this change entirely.

File: package/python-kiwi-spec-template [hunk 6]
```diff
@@ -95,57 +109,74 @@ and cloud systems like Xen, KVM, VMware, EC2 and more.
...
+Requires(post): update-alternatives
+Requires(postun): update-alternatives
+%ifarch x86_64
+Requires:       grub2-x86_64-efi
+%endif
+Requires:       qemu-tools
+Requires:       multipath-tools
 Requires:       squashfs
+Requires:       gptfdisk
+Requires:       zypper
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Zypper is only available in Fedora 26+ and SUSE distributions.

File: package/python-kiwi-spec-template [hunk 7]
```diff
@@ -155,59 +186,73 @@ Python 2 library of the KIWI Image System. Provides an operating system
...
+Requires:         chkconfig
+Requires(post):   chkconfig
+Requires(postun): chkconfig
+Requires:       qemu-img
 Requires:       squashfs-tools
+Requires:       device-mapper-multipath
+Requires:       gdisk
+Requires:       yum
 Provides:       kiwi-packagemanager:yum
+%if 0%{?fedora_version}
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: This will also be in `%{?rhel} >= 8`.

File: package/python-kiwi-spec-template (hunk 3 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: Currently, there are only tags for Fedora 25 in the python-kiwi-boot-packages so I use them also for Fedora 26.

File: package/python-kiwi-spec-template (hunk 5 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: removed this.

File: package/python-kiwi-spec-template (hunk 6 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: So I use it only for SUSE. Whats wrong ?

File: package/python-kiwi-spec-template (hunk 7 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: Is this speculation or a fact that RHEL will switch to dnf for version 8 ?

File: package/python-kiwi-spec-template (hunk 1 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: Removed all redundant License: tags.

File: package/python-kiwi-spec-template (hunk 6 above)
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: It's not wrapped in the appropriate conditional.

If needs `%if 0%{?fedora} >= 26 || 0%{?suse_version}`

File: package/python-kiwi-spec-template [hunk 8]
```diff
@@ -96,56 +108,72 @@ and cloud systems like Xen, KVM, VMware, EC2 and more.
...
+Requires:         chkconfig
+Requires(post):   chkconfig
+Requires(postun): chkconfig
+Requires:       qemu-img
 Requires:       squashfs-tools
+Requires:       device-mapper-multipath
+Requires:       gdisk
+Requires:       yum
 Provides:       kiwi-packagemanager:yum
+%if 0%{?fedora} || 0%{?rhel} > 800
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: It is only  `8`, not `800`.

File: package/python-kiwi-spec-template [hunk 9]
```diff
@@ -155,59 +183,72 @@ Python 2 library of the KIWI Image System. Provides an operating system
...
+Requires:         chkconfig
+Requires(post):   chkconfig
+Requires(postun): chkconfig
+Requires:       qemu-img
 Requires:       squashfs-tools
+Requires:       device-mapper-multipath
+Requires:       gdisk
+Requires:       yum
 Provides:       kiwi-packagemanager:yum
+%if 0%{?fedora} || 0%{?rhel} >= 800
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: It's only `8`, not `800`.

**Comment by mmohring** [comment]
@Conan-Kudo I have now replaced 0{?fedora_version} with 0{?fedora} and 0{?rhel_version} || 0{?centos_version} with 0{?rhel} and put also some 0{?rhel} >= 800 in place for dnf, maybe they use dnf in RHEL 8.

File: package/python-kiwi-spec-template [hunk 10]
```diff
@@ -66,11 +66,16 @@
 %define distro rhel-07.0
 %endif
 
+# Fedora
+# use the rhel templates for CentOS, too
+%if 0%{?fedora_version} >= 25
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Use `0%{?fedora} >= 25` instead.

File: package/python-kiwi-spec-template [hunk 11]
```diff
@@ -79,13 +84,20 @@ Source:         %{name}.tar.gz
 Source1:        %{name}-boot-packages
 Source2:        %{name}-rpmlintrc
 BuildRoot:      %{_tmppath}/%{name}-%{version}-build
-BuildRequires:  python3-devel
-BuildRequires:  python3-setuptools
 BuildRequires:  python-devel
 BuildRequires:  python-setuptools
+%if 0%{?fedora_version} || 0%{?suse_version}
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: `0%{?fedora}` here, please.

File: package/python-kiwi-spec-template (hunk 6 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: Please take a look into the expanded file, the diff fools you. It is wrapped. But what you gonna tell me is that you can use zypper with Fedora 26 ?

Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: Yes, you can. https://admin.fedoraproject.org/pkgdb/package/rpms/zypper/

Comment (mmohring) [inline package/python-kiwi-spec-template]: I fixed it to support zypper also on Fedora >= 26 now.

File: package/python-kiwi-spec-template [hunk 12]
```diff
@@ -96,56 +108,74 @@ and cloud systems like Xen, KVM, VMware, EC2 and more.
 %package -n python2-kiwi
 Summary:        KIWI - Appliance Builder Next Generation
 Group:          Development/Languages/Python
-Provides:       python-kiwi = %{version}-%{release}
-Provides:       kiwi-image:tbz
+Provides:       kiwi
+Obsoletes:      kiwi <= 9.0
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: The `Obsoletes` and `Provides` for `kiwi` need to be removed.

File: package/python-kiwi-spec-template [hunk 13]
```diff
@@ -155,59 +185,74 @@ Python 2 library of the KIWI Image System. Provides an operating system
...
 virtualization and cloud systems like Xen, KVM, VMware, EC2 and more.
 
+%if 0%{?fedora} || 0%{?suse_version}
 # python3-kiwi
 %package -n python3-kiwi
 Summary:        KIWI - Appliance Builder Next Generation
 Group:          Development/Languages/Python
-Provides:       kiwi-image:tbz
+Provides:       kiwi
+Obsoletes:      kiwi <= 9.0
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: The `Obsoletes` and `Provides` for `kiwi` need to be removed.

File: package/python-kiwi-spec-template (hunk 9 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: done all 700 / 800 => 7 / 8, and removed all centos_version now either.

File: package/python-kiwi-spec-template [hunk 14]
```diff
@@ -231,14 +273,13 @@ outside of the scope of kiwi appliance building.
 %ifarch %ix86 x86_64
 %package -n kiwi-pxeboot
 Summary:        KIWI - PXE boot structure
+Group:          System/Management
```
Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: You don't need this, as you get it from the main source package. Same goes for `License` in all subpackages.

Comment (mmohring) [inline package/python-kiwi-spec-template]: I had copied this from the old kiwi package, which has different Group entries for the main package and some of the subpackages.

Comment (Conan-Kudo) [inline package/python-kiwi-spec-template]: If they have different Groups, then it's fine.

File: package/python-kiwi-spec-template [hunk 15]
```diff
@@ -96,56 +108,72 @@ and cloud systems like Xen, KVM, VMware, EC2 and more.
...
 Provides:       kiwi-image:oem
+Provides:       kiwi-image:pxe
+Provides:       kiwi-image:tbz
+Provides:       kiwi-image:vmx
+Provides:       python-kiwi = %{version}-%{release}
+%if 0%{?fedora} || 0%{?suse_version}
 Recommends:     jing
-Requires:       update-alternatives
+%else
+Requires:       jing
```
Comment (schaefi) [inline package/python-kiwi-spec-template]: no requires of jing please, that would mean we require the entire java stack. This should stay as Recommends

Comment (mmohring) [inline package/python-kiwi-spec-template]: How do I make a Recommends with old rhel or centos 7 ?

**Comment by mmohring** [comment]
@schaefi so you want me to redo the patch completely with 3 or 4 simple commits ?

File: package/python-kiwi-spec-template (hunk 15 above)
Comment (schaefi) [inline package/python-kiwi-spec-template]: yes please skip it then. if we run into a validation problem and there is no jing on the system kiwi provides a message to the user that more detailed errors can be obtained if jing would have been installed. imho that's enough information to let a user know what to do. It's a nightmare to require a java based software when it is only an optional thing :) Thanks

**Comment by schaefi** [comment]
I would prefer separate pull requests and handle one thing at a time yes, this also motivates more people to look at it and make it much easier for us to not overlook something.

I'm sorry if this is causing any inconvenience for you. All your changes makes sense and you can be sure we will merge them in, but please let us do it in a more controlled way. Thanks

File: package/python-kiwi-spec-template (hunk 15 above)
Comment (mmohring) [inline package/python-kiwi-spec-template]: OK, I do it. I will now work on 2 or 3 separate pull requests for this patch.

**Comment by mmohring** [comment]
OK, I will do a new version of this patch in separate pull requests.

**Comment by schaefi** [comment]
Thanks much 👍

**Comment by mmohring** [comment]
@schaefi I have now created a simplified patch with only adding the CentOS / RHEL / Fedora support. Its pull request #439 .
